### PR TITLE
OSSM-3802 Implement SMCP.spec.meshConfig.discoverySelectors

### PR DESCRIPTION
**Pull request by mayleighnmyers**
https://issues.redhat.com/browse/OSSM-3802

This PR allows us to make discoverySelectors GA

### CODE REVIEW

File: pkg/apis/maistra/v2/meshconfig.go
Comment (jewertow) [inline pkg/apis/maistra/v2/meshconfig.go]:
```suggestion
	DiscoverySelectors []*v1.LabelSelector `json:"discoverySelectors,omitempty"`
```

File: resources/helm/v2.4/istio-control/istio-discovery/files/gateway-injection-template.yaml
Comment (jewertow) [inline resources/helm/v2.4/istio-control/istio-discovery/files/gateway-injection-template.yaml]: Please revert this change.

**Review by jewertow — CHANGES_REQUESTED** [review]
This PR lacks an implementation of v1 / v2 API conversion.

File: pkg/apis/maistra/conversion/discovery_selectors_test.go
Comment (luksa) [inline pkg/apis/maistra/conversion/discovery_selectors_test.go:116]: Looks like you lost the additional test that you added after our meeting.

**Review by luksa — CHANGES_REQUESTED** [review]

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (luksa) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: You need to remove this if statement (just this line, not the entire if block) to make the `empty` test pass.

**Comment by luksa** [comment]
@mayleighnmyers FYI I ran `make gen` for you and commited the changes to your PR, so be sure to pull them before you continue.

File: pkg/apis/maistra/conversion/discovery_selectors_test.go
Comment (jewertow) [inline pkg/apis/maistra/conversion/discovery_selectors_test.go]:
```suggestion
							MatchExpressions: []metav1.LabelSelectorRequirement{
```

File: pkg/apis/maistra/conversion/discovery_selectors_test.go
Comment (jewertow) [inline pkg/apis/maistra/conversion/discovery_selectors_test.go]:
```suggestion
	}
```

File: pkg/apis/maistra/conversion/discovery_selectors_test.go
Comment (jewertow) [inline pkg/apis/maistra/conversion/discovery_selectors_test.go]:
```suggestion
									Operator: metav1.LabelSelectorOpIn,
```

**Review by jewertow — CHANGES_REQUESTED** [review]
Tests fail on compilation.

**Review by luksa — COMMENTED** [review]
LGTM, but I'd like @jewertow to double-check.

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (jewertow) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: You can simplify these conditions as below:
```suggestion
	if discoverySelectors, err := sliceToValues(untypedSlice); err == nil {
		return setHelmValue(out, "meshConfig.discoverySelectors", discoverySelectors)
	} else {
		return err
	}
```

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (jewertow) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: You can remove this boolean variable and simplify function as below:
```suggestion
        var discoverySelectors []*metav1.LabelSelector
	if ds, ok, err := in.GetAndRemoveSlice("meshConfig.discoverySelectors"); ok {
		if err := fromValues(ds, &discoverySelectors); err != nil {
			return err
		}
		if out.MeshConfig == nil {
			out.MeshConfig = &v2.MeshConfig{}
		}
		out.MeshConfig.DiscoverySelectors = discoverySelectors
	} else if err != nil {
		return err
	}
```

**Review by jewertow — CHANGES_REQUESTED** [review]
Two blockers:
1. Functions `populateDiscoverySelectorsValues` and `populateDiscoverySelectorsConfig` are not used.
2. We need validation that will allow to set `discoverySelectors` only in cluster-wide mode. Look at this PR as reference: https://github.com/maistra/istio-operator/pull/1143.

File: pkg/apis/maistra/conversion/conversion_v1_to_v2.go
Comment (luksa) [inline pkg/apis/maistra/conversion/conversion_v1_to_v2.go]: You can merge this if statement with the one for extension providers. No need to have two if statements that check the version.

File: pkg/apis/maistra/conversion/conversion_v2_to_v1.go
Comment (luksa) [inline pkg/apis/maistra/conversion/conversion_v2_to_v1.go]: Same here.

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (luksa) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: This empty line is not needed.

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (luksa) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: This empty line is not needed.

File: pkg/apis/maistra/conversion/discovery_selectors.go
Comment (luksa) [inline pkg/apis/maistra/conversion/discovery_selectors.go]: You need to `return nil` in this line. This will resolve the syntax error you're probably seeing here or in the next line.

**Comment by luksa** [comment]
@mayleighnmyers You still need to implement #2 in @jewertow 's comment: 

> We need validation that will allow to set discoverySelectors only in cluster-wide mode. Look at this PR as reference: https://github.com/maistra/istio-operator/pull/1143.

**Comment by luksa** [comment]
@mayleighnmyers You need to rebase to upstream/maistra-2.4 to get Jacek's changes from #1143 in. Then add a similar check that he made for extensionProviders, but for discoverySelectors. The reason you don't see his code is because you created your branch before his changes were merged.

To rebase, run the following command: `git pull --rebase upstream maistra-2.4`

**Comment by jewertow** [comment]
Please reset to 10f45c7af0b395a0adaef96b0bdd1beccb6b2ad9 and then rebase.

File: pkg/controller/versions/util.go
Comment (jewertow) [inline pkg/controller/versions/util.go]:
```suggestion
		return append(allErrors, fmt.Errorf("the spec.meshConfig.discoverySelectors field is only supported in version 2.4 and above"))
```

**Comment by luksa** [comment]
@mayleighnmyers I squashed all the commits and rebased to `maistra-2.4`, so this PR is now easier to work with.

File: pkg/controller/versions/util.go
Comment (luksa) [inline pkg/controller/versions/util.go]: No, this function is checking the wrong thing. It needs to return an error only when `mode` is `MultiTenant` and `discoverySelectors` are non-nil. If `mode` is `ClusterWide`, then `discoverySelectors` *may* be set (but don't need to be).

Comment (mayleighnmyers) [reply pkg/controller/versions/util.go]: So would I change ClusterWide to MultiTenant?

Comment (luksa) [reply pkg/controller/versions/util.go]: I'd simply name the function `checkDiscoverySelectors`. As for the actual check: it's not really simple, as checking `spec.Mode` is not enough, because `mode` can be set in an SMCP profile/template (i.e. `mode` might be nil in the `spec`). But we probably have similar bugs elsewhere. 

@jewertow Why do we need to disallow the use of discoverySelectors in MultiTenant mode? Even if we use xns-informer, the selectors would still work, wouldn't they? Also, we allowed users to use `techPreview.meshConfig.discoverySelectors`, did we not?

If anything, we should disallow the use of `techPreview.meshConfig.discoverySelectors` and tell users to use `meshConfig.discoverySelectors` instead.

Comment (mayleighnmyers) [reply pkg/controller/versions/util.go]: @luksa So I changed the name of the function, but a little confused on what I should do for the `spec.mode` or is this something that will eventually be changed later since we might have it used this way other places?

Comment (luksa) [reply pkg/controller/versions/util.go]: @mayleighnmyers let's wait for @jewertow to explain why we shouldn't allow discoverySelectors in multi-tenant mode. If it turns out we can allow them, we can just remove the function altogether.

Comment (luksa) [reply pkg/controller/versions/util.go]: Actually, it would be great if you could try this out (set mode: MultiTenant and specify spec.meshConfig.discoverySelectors). The best way would be to add another test case to maistra-test-tool (like the one in https://github.com/maistra/maistra-test-tool/pull/519).

Comment (luksa) [reply pkg/controller/versions/util.go]: I tested MultiTenant with discoverySelectors and realized we probably don't want to allow them, as Jacek suggested. So I implemented this in the validating webhook.

@jewertow Can you review this, so we can get it merged?

**Review by luksa — APPROVED** [review]

**Review by jewertow — APPROVED** [review]

**Pull request merged**
